File: Assets/Scripts/Scene/Level/XLevelSpawnMgr.cs
using UnityEngine;
using System.IO;

namespace Level
{
    public class XLevelSpawnMgr : XSingleton<XLevelSpawnMgr>
    {
        private XLevelSpawnInfo _curSpawner;
        private float _time = 0;
        public bool BossExtarScriptExecuting = false;
        public bool IsCurrentLevelWin { get; set; }
        public bool IsCurrentLevelFinished { get; set; }

        public XLevelSpawnInfo currSpawn { get { return _curSpawner; } }

        public void Update(float deltaT)
        {
            if (IsCurrentLevelFinished) return;
            _time += deltaT;
            if (_curSpawner != null) _curSpawner.Update(_time);
        }

        public void ForceLevelFinish(bool win)
        {
            IsCurrentLevelFinished = true;
            if (win)
            {
                XLevelState ls = XLevelStatistics.singleton.ls;
                IsCurrentLevelWin = true;
                OnLevelFinish(ls._lastDieEntityPos + new Vector3(0.0f, ls._lastDieEntityHeight, 0.0f) / 2, ls._lastDieEntityPos, 500, 0, true);
            }
            else
            {
                OnLevelFailed();
            }
        }

        public void OnLevelFinish(Vector3 dropInitPos, Vector3 dropGounrdPos, uint money, uint itemCount, bool bKillOpponent)
        {
        }

        public void OnLevelFailed()
        {
        }

        public void OnEnterScene(uint sceneid)
        {
            _time = 0;
            XLevelScriptMgr.singleton.CommandCount = 0;
            string configFile = XScene.singleton.SceneRow.SceneFile;
            if (configFile.Length == 0)
            {
                _curSpawner = null;
                XLevelScriptMgr.singleton.ClearWallInfo();
                XLevelScriptMgr.singleton.Reset();
                return;
            }

            if (_curSpawner == null) _curSpawner = new XLevelSpawnInfo();
            else _curSpawner.Clear();
            Stream s = XResources.ReadText("Table/" + configFile);
            //using (StreamReader sr = new StreamReader(s))
            StreamReader sr = new StreamReader(s);
            {
                string line = sr.ReadLine();
                int totalWave = int.Parse(line);
                line = sr.ReadLine();
                int PreloadWave = int.Parse(line);
                for (int i = 0; i < PreloadWave; i++)
                {
                    line = sr.ReadLine();
                    string[] info = line.Split(',');
                    int enemyID = int.Parse(info[0].Substring(3));
                    int count = int.Parse(info[1]);
                    _curSpawner.preloadInfo.Add(enemyID, count);
                }
                for (int id = 0; id < totalWave; id++)
                {
                    XLevelWave _wave = new XLevelWave();
                    _wave.ReadFromFile(sr);
                    _curSpawner.waves.Add(_wave);

                    XLevelDynamicInfo dInfo = new XLevelDynamicInfo();
                    dInfo.id = _wave.ID;
                    dInfo.totalCount = _wave.monsterPos.Count + _wave.RoundCount;
                    dInfo.Reset();
                    _curSpawner.wavesDynamicInfo.Add(_wave.ID, dInfo);
                }

                XResources.ClearStream(s);
            }
        }
    }
}

File: tools_proj/XCPP/GameCore/XLevelSpawnMgr.cpp
#include "XLevelSpawnMgr.h"
#include "XEntityStatistics.h"
#include "XScene.h"
#include <assert.h>

extern std::string UNITY_STREAM_PATH;

XLevelSpawnMgr::XLevelSpawnMgr()
{
}

XLevelSpawnMgr::~XLevelSpawnMgr()
{
	ClearAll();
}

void XLevelSpawnMgr::ClearAll()
{
	delete _curSpawner;
	_curSpawner = NULL;
	for (std::map<uint, std::map<int, XLevelWave*>>::iterator itr = m_StaticWaves.begin();
		itr != m_StaticWaves.end(); itr++)
	{
		for (size_t i = 0; i < itr->second.size(); i++)
		{
			delete itr->second[i];
		}
		itr->second.clear();
	}
	m_StaticWaves.clear();
}

void XLevelSpawnMgr::OnEnterScene(uint sceneid)
{
	_time = 0;
	XLevelScriptMgr::Instance()->CommandCount = 0;
	SceneListRow* row = XScene::Instance()->getSceneRow();
	char* configFile = row->scenefile;
	if (configFile == NULL)
	{
		delete _curSpawner;
		_curSpawner = NULL;
		XLevelScriptMgr::Instance()->ClearWallInfo();
		XLevelScriptMgr::Instance()->Reset();
		return;
	}
	if (_curSpawner == NULL) _curSpawner = new XLevelSpawn();
	else _curSpawner->Clear();
	ifstream infile;
	std::string path = UNITY_STREAM_PATH + tostring(configFile)+".txt";
	infile.open(path.data(), ios::in);
	assert(infile.is_open());
	string line;
	getline(infile, line);
	int totalWave = atoi(line.c_str());
	getline(infile, line);
	int preloadwave = atoi(line.c_str());
	for (int i = 0; i < preloadwave; i++)
	{
		getline(infile, line);
		std::vector<string> info = split(line, ',');
		int enemyID = atoi(info[0].substr(3).c_str());
		int count = atoi(info[1].c_str());
		_curSpawner->preloadInfo.insert(std::make_pair(enemyID, count));
	}
	for (int i = 0; i < totalWave; i++)
	{
		XLevelWave* wave = new XLevelWave();
		wave->ReadFromFile(infile);
		_curSpawner->waves.push_back(wave);

		XLevelDynamicInfo* dInfo = new XLevelDynamicInfo();
		dInfo->id = wave->ID;
		dInfo->totalCount = wave->count;
		dInfo->Reset();
		_curSpawner->wavesDynamicInfo.insert(std::make_pair(wave->ID, dInfo));
	}
	infile.close();
}


void XLevelSpawnMgr::Update(float deltaT)
{
	if (IsCurrentLevelFinished) return;
	_time += deltaT;
	if (_curSpawner) _curSpawner->Update(_time);
}

void XLevelSpawnMgr::ForceLevelFinish(bool win)
{
	IsCurrentLevelFinished = true;
	if (win)
	{
		XLevelState* ls = XLevelStatistics::Instance()->ls;
		IsCurrentLevelWin = true;
		Vector3 v3 = Vector3(0.0f, ls->_lastDieEntityHeight*0.5f, 0.0f);
		OnLevelFinish(ls->_lastDieEntityPos + v3, ls->_lastDieEntityPos, 500, 0, true);
	}
	else
	{
		OnLevelFailed();
	}
}

void XLevelSpawnMgr::OnLevelFinish(Vector3 dropInitPos, Vector3 dropGounrdPos, uint money, uint itemCount, bool bKillOpponent)
{
	LOG("LEVEL FINISH");
}

void XLevelSpawnMgr::OnLevelFailed()
{
	LOG("LEVEL FAILED");
}
